### PR TITLE
common: fix that $host always expands to localhost instead of actual hostname

### PR DESCRIPTION
**Pull request by Liuchang0812**
* add function : ceph_get_hostname, ceph_get_short_hostname
* add unittest for hostname function
* expand $host in config.cc

TODO: update documentation about this feature

Fixes: http://tracker.ceph.com/issues/11081

Signed-off-by: liuchang0812 <liuchang0812@gmail.com>

### CODE REVIEW

**Comment by Liuchang0812** [comment]
The following tests FAILED:
	 69 - unittest_config (Failed)
Errors while running CTest
Build step 'Execute shell' marked build as failure

**Comment by Liuchang0812** [comment]
```
/home/liuchang/WorkSpace/ceph/src/test/common/test_config.cc:69: Failure
      Expected: public_network + " " + public_network + " " + lockdep + " " + "localhost"
      Which is: "NETWORK NETWORK true localhost"
To be equal to: val
      Which is: "NETWORK NETWORK true liuchang-VirtualBox"
```

**Comment by Liuchang0812** [comment]
I have tested this in my pc. It works ok.

```
# ceph.conf
[osd.0]                                                                                                             
        osd crush location = host=${host}lc root=not-default                     
[osd.1]                                                                                                                  
        osd crush location = host=lc${host}lc root=not-default                   
[osd.2]                                                                          
        host = osd2                                                              
        osd crush location = host=${host} root=not-default  

# query crush location 
➜  build git:(wip-11081) ✗ ./bin/ceph-conf --name=osd.0 --lookup=osd_crush_location
host=liuchang-VirtualBoxlc root=not-default
➜  build git:(wip-11081) ✗ ./bin/ceph-conf --name=osd.1 --lookup=osd_crush_location
host=lcliuchang-VirtualBoxlc root=not-default
➜  build git:(wip-11081) ✗ ./bin/ceph-conf --name=osd.2 --lookup=osd_crush_location
host=osd2 root=not-default
```

@liewegas @tchaikov mind taking a look. very thanks

File: src/common/config_opts.h [hunk 1]
```diff
@@ -13,7 +13,7 @@
  */
 
 /* note: no header guard */
-OPTION(host, OPT_STR, "localhost")
+OPTION(host, OPT_STR, "default")  // "default" means that host will be `hostname -s`
```
Comment (liewegas) [inline src/common/config_opts.h]: can we make it $hostname and $fqdn perhaps?  if it's magic it should be a normal substitution variable..

Comment (liewegas) [inline src/common/config_opts.h]: hrm, maybe what we should really do is eliminate 'host' as a config option and turn it into a proper meta.  i'm not sure why users would be configuring this?

Comment (liewegas) [inline src/common/config_opts.h]: if we think it is important, than we should make "" the value that looks at the real hostname.  meaning the config option is nothing more than an override

Comment (Liuchang0812) [inline src/common/config_opts.h]: hi, @liewegas thanks for your comments.  I'm agree with your advice that use $hostname and $fqdn.

but, there are a lot of $host config in user product env probably. Eliminating $host must be carefully.

in RGW, there is a option called `rgw dns name`, that is used in s3 authorization.In my dev/test env,  I would like to config this option to really hostname. so providing a $hostname is useful.

Comment (liewegas) [inline src/common/config_opts.h]: I think if we just change teh default from "localhost" to "", and change "" to fill in the real hostname, we'll be pretty safe:

- if someone configured a host, that won't change
- if someone is using $host but is getting the default of 'localhost', they probably haven't noticed.. and hopefully won't notice if it starts getting a real hostname.  We can add something to the release notes.
- Same for if 'host = ' and they are using $host

Comment (Liuchang0812) [inline src/common/config_opts.h]: OK! sounds perfect, I am updating it. 

so, we don't need to provide ${hostname} and ${fdqn} now?

**Comment by liewegas** [comment]
hmm, i don't think so.  let's keep it simple.

**Comment by Liuchang0812** [comment]
@liewegas OK!

**Comment by Liuchang0812** [comment]
@liewegas updated! take a look please.

**Comment by Liuchang0812** [comment]
Ping! Please take a look, I will have a long holiday, i want to get it done before the holiday. Very thanks

**Comment by Liuchang0812** [comment]
@liewegas @tchaikov ping

**Pull request merged**